smarts/zoo/agent_spec.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (C) 2021. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import inspect
+import logging
+import warnings
+from dataclasses import dataclass
+from typing import Any, Callable, Optional
+
+import cloudpickle
+
+warnings.simplefilter("once")
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class AgentSpec:
+    """A configuration that is used by SMARTS environments.
+
+    .. code-block:: python
+
+        agent_spec = AgentSpec(
+            interface=AgentInterface.from_type(AgentType.Laner),
+            agent_params={"agent_function": lambda _: "keep_lane"},
+            agent_builder=Agent.from_function,
+        )
+
+        env = gym.make(
+            "smarts.env:hiway-v0",
+            scenarios=["scenarios/loop"],
+            agent_specs={agent_id: agent_spec},
+        )
+
+        agent = agent_spec.build_agent()
+
+    Refer to the Agent documentation.
+    """
+
+    from smarts.core.agent import Agent
+    from smarts.core.agent_interface import AgentInterface
```

Comment:
Why are the imports inside the class and can they be moved out?

Reply:
This was previously needed because of the approach we used for aliasing the deprecated class which resulted in circular imports. It appears we don't have that issue anymore so I've moved it out again.